exporter/loadbalancingexporter/resolver_k8s.go

```diff
@@ -0,0 +1,225 @@
+		port:           ports,
+		once:           &sync.Once{},
+		endpointsStore: epsStore,
+		epsListWatcher: epsListWatcher,
+		handler:        h,
+		stopCh:         make(chan struct{}),
+	}
+	h.callback = r.resolve
+
+	return r, nil
+}
+
+func (r *k8sResolver) start(ctx context.Context) error {
+	if r.once == nil {
+		return fmt.Errorf("cannot Start() partial k8sResolver (nil *sync.Once)")
+	}
+	if r.handler == nil {
+		return fmt.Errorf("cannot Start() partial k8sResolver (nil *handler)")
+	}
+	var initErr error
+	r.once.Do(func() {
+		if r.epsListWatcher != nil {
+			r.logger.Debug("creating and starting endpoints informer")
+			epsInformer := cache.NewSharedInformer(r.epsListWatcher, &corev1.Endpoints{}, 0)
+			if _, err := epsInformer.AddEventHandler(r.handler); err != nil {
+				r.logger.Error("error adding event handler to endpoints informer", zap.Error(err))
+			}
+			go epsInformer.Run(r.stopCh)
+			if !cache.WaitForCacheSync(r.stopCh, epsInformer.HasSynced) {
+				r.logger.Error("error endpoints informer not sync")
+				initErr = errors.New("endpoints informer not sync")
+				return
+			}
+		}
+	})
+	if initErr != nil {
+		return initErr
+	}
+
+	r.logger.Debug("K8s service resolver started",
+		zap.String("service", r.svcName),
+		zap.String("namespace", r.svcNs),
+		zap.Int32s("ports", r.port))
+	return nil
+}
+
+func (r *k8sResolver) shutdown(ctx context.Context) error {
+	r.changeCallbackLock.Lock()
+	r.onChangeCallbacks = nil
+	r.changeCallbackLock.Unlock()
+
+	close(r.stopCh)
+	r.shutdownWg.Wait()
+	return nil
+}
+func newInClusterClient() (kubernetes.Interface, error) {
+	cfg, err := config.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	return kubernetes.NewForConfig(cfg)
+}
+
+func (r *k8sResolver) resolve(ctx context.Context) ([]string, error) {
+	r.shutdownWg.Add(1)
+	defer r.shutdownWg.Done()
+
+	var backends []string
+	r.endpointsStore.Range(func(address, value any) bool {
+		addr := address.(string)
+		if len(r.port) == 0 {
+			backends = append(backends, addr)
+		} else {
+			for _, port := range r.port {
+				backends = append(backends, fmt.Sprintf("%s:%d", addr, port))
+			}
+		}
+		return true
+	})
+
+	// keep it always in the same order
+	sort.Strings(backends)
+
+	if equalStringSlice(r.endpoints, backends) {
+		return r.endpoints, nil
+	}
+
+	// the list has changed!
+	r.updateLock.Lock()
+	r.endpoints = backends
+	r.updateLock.Unlock()
+	_ = stats.RecordWithTags(ctx, resolverSuccessTrueMutators, mNumBackends.M(int64(len(backends))))
+
+	// propagate the change
+	r.changeCallbackLock.RLock()
+	for _, callback := range r.onChangeCallbacks {
+		callback(r.endpoints)
+	}
+	r.changeCallbackLock.RUnlock()
+	return r.endpoints, nil
+}
+
+func (r *k8sResolver) onChange(f func([]string)) {
+	r.changeCallbackLock.Lock()
+	defer r.changeCallbackLock.Unlock()
+	r.onChangeCallbacks = append(r.onChangeCallbacks, f)
+}
+
+const inClusterNamespacePath = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
```

Comment:
Is this the namespace for the current pod, or is it for the service account being used to run this pod? Are the namespaces always the same, or can a service account from one space be used on a pod on a different namespace?

Reply:
serviceaccount is a resource under a namespace, which in this case is the namespace where the current pod is located.